### DISCUSSION TITLE
Policy enumeration for a generator?

### DISCUSSION THREAD

**Discussion by acicali** [ImageMagick/ImageMagick#3314] [Development]
Hi,

I'd like to configure ImageMagick to have the least possible privileges to suit my needs. To aid myself and others in this goal, it'd be great if there were a frontend for this that took the guesswork out of the process - like a policy.xml generator.

I'm unable to find such a policy generator on the web, so I thought perhaps I'd create one. For that to work, I'd prefer to be able to somehow enumerate the available policies that ImageMagick provides, rather than work off of a static list that will change over time.

Is there a known method for enumerating ImageMagick policies? I don't want to grep the source code.

Thanks for reading.

**Comment by snibgo** [comment]
How about:
```
f:\web\im>%IMG7%magick -list PolicyDomain
Cache
Coder
Delegate
Filter
Module
Path
Resource
System

f:\web\im>%IMG7%magick -list Resource
Resource limits:
  Width: 107.374MP
  Height: 107.374MP
  Area: 6.40681GP
  List length: unlimited
  Memory: 2.9834GiB
  Map: 5.9668GiB
  Disk: unlimited
  File: 1536
  Thread: 8
  Throttle: 0
  Time: unlimited
```

**Comment by acicali** [comment]
Thanks @snibgo. Admittedly, I'm not very experienced with ImageMagick, and a complete novice to its policy.xml. From what I've seen so far, policy.xml is a fairly simple structure. Each <policy> tag contains a single domain attribute, plus either "name/value" attributes or "rights/pattern" attributes. The optional "stealth" attribute seems to hide the property from "_convert -list policy_" calls.

- The "_convert -list PolicyDomain_" command does indeed seem to provide all of the possible domains that can be configured within the policy.
- The "_convert -list Resource_" command seems to provide the resource domain properties, but the docs (https://imagemagick.org/script/security-policy.php) also refer to a "temporary-path" value, so perhaps it's not exhaustive.

Other domains don't seem to offer the same convenience. For instance "-list Path" or "-list System" do not work. Not sure yet how any of this translates to an actual policy document, but I'll tinker a bit and see what I can come up with.

Thanks again!

**Comment by snibgo** [comment]
For rights applicable to paths, see "magick -list PolicyRights". For example, policy.xml might contain:
```
  <policy domain="path" rights="none" pattern="*snib*" />
```
This would prevent me from reading or writing any files that match \*snib\*. There is no fixed list of possible patterns, of course.

Internally, "temporary-path" is not treated in the same way as other "Resource" limits. See policy.c.
